cli/root.go

```diff
@@ -1,83 +1,177 @@
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			panic(err)
+		}
+		var config map[string]string
+		err = yaml.Unmarshal(b, &config)
+		if err != nil {
+			panic(err)
+		}
+		resolver.configData = append(resolver.configData, config)
+	}
+
+	return resolver
+}
+
+func (r amtoolConfigResolver) Resolve(key string, context *kingpin.ParseContext) ([]string, error) {
+	for _, c := range r.configData {
+		if v, ok := c[key]; ok {
+			return []string{v}, nil
+		}
+	}
+	return nil, nil
+}
 
-The alertmanager tool will read a config file from the --config cli argument, AMTOOL_CONFIG environment variable or
-from one of two default config locations. Valid config file formats are JSON, TOML, YAML, HCL and Java Properties, use
-whatever makes sense for your project.
+// This function maps things which have previously had different names in the
+// config file to their new names, so old configrations keep working
+func backwardsCompatibilityResolver(key string) string {
+	switch key {
+	case "require-comment":
+		return "comment_required"
+	}
+	return key
+}
 
-The default config file paths are $HOME/.config/amtool/config.yml or /etc/amtool/config.yml
+func init() {
+	longHelpText["root"] = `View and modify the current Alertmanager state.
+
+Config File:
+The alertmanager tool will read a config file in YAML format from one of two
+default config locations: $HOME/.config/amtool/config.yml or
+/etc/amtool/config.yml
 
-The accepted config options are as follows:
+All flags can be given in the config file, but the following are the suited for
+static configuration:
 
 	alertmanager.url
 		Set a default alertmanager url for each request
 
 	author
-		Set a default author value for new silences. If this argument is not specified then the username will be used
+		Set a default author value for new silences. If this argument is not
+		specified then the username will be used
 
-	comment_required
-		Require a comment on silence creation
+	require-comment
+		Bool, whether to require a comment on silence creation. Defaults to true
 
 	output
 		Set a default output type. Options are (simple, extended, json)
-	`,
+
+	date.format
+		Sets the output format for dates. Defaults to "2006-01-02 15:04:05 MST"
+`
 }
 
-// Execute adds all child commands to the root command sets flags appropriately.
-// This is called by main.main(). It only needs to happen once to the rootCmd.
+// Execute parses the arguments and executes the corresponding command, it is
+// called by cmd/amtool/main.main().
 func Execute() {
-	if err := RootCmd.Execute(); err != nil {
-		os.Exit(1)
+	format.InitFormatFlags(app)
+
+	app.Version(version.Print("amtool"))
+	app.GetFlag("help").Short('h')
+	app.UsageTemplate(kingpin.CompactUsageTemplate)
+	app.Flag("long-help", "Give more detailed help output").UsageAction(&kingpin.UsageContext{
+		Template: longHelpTemplate,
+		Vars:     map[string]interface{}{"LongHelp": longHelpText},
+	}).Bool()
+
+	configResolver := newConfigResolver()
```

Comment:
I would prefer to have `newConfigResolver` return an error, and instead of panicking on a file read error or a yaml unmarshal error, provide a shorter and clearer message. `panic()`'s stack trace probably won't aid a user much, and generate a github issue when they just need to fix their yaml.

Reply:
Agreed.